### PR TITLE
feat!: fees distribution

### PR DESCRIPTION
**Pull request by jawid-h**
## Issue being fixed or feature implemented

Introducing the concept of `Fee Pools` into the project, to be able to process `Delayed Reward Distribution`.

### Delayed Reward Distribution (from the document)
Neither processing fees nor storage fees are immediately rewarded to validators. Instead, processing fees are rewarded in the following Epoch, and storage fees are rewarded to every Epoch of the storage lifetime (50 years for permanent storage), starting with the following Epoch. Delaying distribution has two main objectives: first, it is designed to prevent masternodes from increasing load on this system when they know they are part of the validator set. Second, masternodes incur storage costs for the lifetime of the project. If storage costs were distributed immediately, it could become cost prohibitive to start new masternodes at a future date if the ongoing storage costs could not be offset from new storage fees.

## What was done?

- introduced process_block_fees function
- introduced `init_chain`, `block_begin` and `block_end` ABCI handlers
- moved the encode functions from contract to common module
- proposed a new batch structure to simplify the logic

## How Has This Been Tested?

- integration tests (from smaller parts to main flow functions)

## Breaking Changes

- introduced concept of `ABCI` function to the `rs-drive` awaits `JS` part to use it's respective functions  (e.g. `init_chain`) to achieve `RootTree` creation now

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: drive/src/drive/fee_pools.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+use crate::drive::Drive;
+use crate::error::Error;
+use crate::fee::pools::fee_pools::FeePools;
+
+use chrono::Utc;
+
+impl Drive {
+    pub fn init_fee_pools(
+        &self,
+        genesis_time: i64,
```
Comment (shumkov) [inline drive/src/drive/fee_pools.rs]: It's not used

File: drive/src/fee/pools/epoch/proposers.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            .get(
+                self.get_path(),
+                constants::KEY_FIRST_PROPOSER_BLOCK_HEIGHT.as_bytes(),
+                transaction,
+            )
+            .map_err(Error::GroveDB)?;
+
+        if let Element::Item(item) = element {
+            Ok(u64::from_le_bytes(
+                item.as_slice().try_into().expect("invalid item length"),
```
Comment (shumkov) [inline drive/src/fee/pools/epoch/proposers.rs]: I guess you need to map it to Drive error too. In other places too.

File: drive/src/fee/pools/fee_pools.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        Ok(())
+    }
+
+    pub fn get_oldest_epoch_pool(
+        &self,
+        epoch_index: u16,
+        transaction: TransactionArg,
+    ) -> Result<EpochPool, Error> {
+        if epoch_index == 1 {
```
Comment (shumkov) [inline drive/src/fee/pools/fee_pools.rs]: I guess the oldest can be an epoch 0 for epoch 1. Isn't it?

**Review by shumkov — DISMISSED** [review]

File: drive/Cargo.toml [hunk 4]
```diff
@@ -19,12 +19,15 @@ hex = "0.4.3"
 array_tool = "1.0.3"
 rand_distr = "0.4.3"
 tempfile = "3"
-serde_json = "1.0.73"
+serde_json = { version="1.0", features=["preserve_order"] }
```
Comment (antouhou) [inline drive/Cargo.toml:22]: What's exactly the reason behind dropping the patch version of serde?

Comment (shumkov) [reply drive/Cargo.toml:22]: We just copied from rust dpp Cargo.toml

**Pull request merged**